repair/row_level.cc

```diff
@@ -53,6 +53,82 @@
 
 extern logging::logger rlogger;
 
+class token_bucket {
+private:
+    size_t _fill_rate;
+    size_t _max_tokens;
+    size_t _current_tokens;
+    seastar::condition_variable _cond;
+    seastar::timer<std::chrono::steady_clock> _fill_timer;
+    std::chrono::milliseconds _fill_interval{10};
+private:
+    void fill() {
+        auto tokens = _fill_rate * _fill_interval.count() / std::chrono::milliseconds(1000).count();
+        _current_tokens = std::min(_max_tokens, _current_tokens + tokens);
+        _cond.broadcast();
+    }
+    bool enabled() {
+        return _fill_rate != 0;
+    }
+public:
+    token_bucket(size_t fill_rate, size_t max_tokens)
+        : _fill_rate(fill_rate)
+        , _max_tokens(max_tokens)
+        , _current_tokens(max_tokens) {
+        _fill_timer.set_callback([this] { fill(); });
+        set_rate(fill_rate);
+    }
+    future<> consume(size_t tokens) {
+        if (!enabled()) {
+            return make_ready_future<>();
+        }
+        if (tokens > _max_tokens) {
+            rlogger.warn("token_bucket: tokens={} requested are bigger than max_tokens={}", tokens, _max_tokens);
+            return make_ready_future<>();
```

Comment:
Wouldn't this allow very large rows to just bypass the limit?

Reply:
It is better to bypass than hang forever, no?

Reply:
Wouldn't it be better to clamp `tokens` to `[0, _max_tokens]`? This way we wouldn't completely bypass the limitation, but still wouldn't block.

Reply:
I think it's going to be ok, @denesb.
Note that in that case a `_current_tokens` would go negative and it would take a few quantums for it to recover. So, in the average the limit will be fulfilled. While momentarily it would be higher.